### PR TITLE
[Uniswap V3 Merge] aka the worst thing ever

### PR DESCRIPTION
**Pull request by W3stside**
Merges in the latest from Uniswap V3

TODO:
1. Review TODO comments
2. Check package scripts - especially abi building (v3) and lingui translation compilation
3. Styles are kind of effed
  a. Font-family? Where is it set as it's broken in app rn (styles related)
4. test all flows:
- [ ] can trade
- [ ] can wrap
- [ ] can delete order
- [ ] toast notifications correct
- [ ] sell orders
- [ ] buy orders
- [ ] usd estimations are correct
- [ ] trade numbers are correct

## Testing
The changes include a new way to start the app:
// lingui translation files compilation
1. ~`yarn i18n:extract`~
2. `yarn start`

## Mod changelog
1. AccountDetailsMod
2. ButtonMod
3. TransactionMod
4. CurrencyInputPanelMod
5. MenuMod
6. PopupItemMod
7. TransactionPopupMod
8. CurrencyListMod
9. ImportListMod
10. ManageListsMod
11. SettingsMod
12. AdvancedSwapDetailsMod
13. AdvancedSwapDetailsDropdownMod
14. ConfirmSwapModalMod
15. SwapModalFooterMod
16. SwapModalHeaderMod
17. UnsupportedCurrencyFooterMod
18. TooltipMod
19. TransactionConfirmationModalMod
20. TransactionSettingsMod
21. QuestionHelperMod
22. Web3StatusMod
23. Constants/tokens

## Notes
CrowdIn translations:
Help CowSwap reach a global audience! 

CowSwap uses [Crowdin](https://crowdin.com/project/uniswap-interface) 
for managing translations. Whenever a new string is added to the project,
it gets uploaded to Crowdin for translation by [this workflow](./.github/workflows/crowdin.yaml).

Every hour, translations are synced from Crowdin in [this other workflow](./.github/workflows/crowdin-sync.yaml).

You can contribute by joining Crowdin to proofread existing translations [here](https://crowdin.com/project/uniswap-interface/invite?d=93i5n413q403t4g473p443o4c3t2g3s21343u2c3n403l4b3v2735353i4g4k4l4g453j4g4o4j4e4k4b323l4a3h463s4g453q443m4e3t2b303s2a35353l403o443v293e303k4g4n4r4g483i4g4r4j4e4o473i5n4a3t463t4o4)

Or, ask to join us as a translator in the Discord!

### CODE REVIEW

File: src/components/AccountDetails/Copy.tsx [hunk 1]
```diff
@@ -1,9 +1,10 @@
 import React from 'react'
-import styled from 'styled-components'
+import styled from 'styled-components/macro'
```
Comment (alfetopito) [inline src/components/AccountDetails/Copy.tsx:2]: Have you found out why is this imported from `/macro` ?

File: src/components/AddressInputPanel/index.tsx [hunk 2]
```diff
@@ -102,8 +102,11 @@ export default function AddressInputPanel({
                 Recipient
               </TYPE.black>
               {address && chainId && (
-                <ExternalLink href={getEtherscanLink(chainId, name ?? address, 'address')} style={{ fontSize: '14px' }}>
-                  (View on Etherscan)
+                <ExternalLink
+                  href={getExplorerLink(chainId, name ?? address, ExplorerDataType.ADDRESS)}
+                  style={{ fontSize: '14px' }}
+                >
+                  (View on Explorer)
```
Comment (alfetopito) [inline src/components/AddressInputPanel/index.tsx:109]: Oopsie, Uni team missed on translate spot.
Unless it's inside `ExternalLink` component, which I didn't get to yet...

(No action to take, just a comment)

File: src/components/CurrencyInputPanel/FiatValue.tsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        <Trans>
+          ~$ <HoverInlineText text={fiatValue?.toSignificant(6, { groupSeparator: ',' })} />
+        </Trans>
+      ) : (
+        ''
+      )}
+      {priceImpact ? (
+        <span style={{ color: priceImpactColor }}>
+          {' '}
+          (<Trans>{priceImpact.multiply(-1).toSignificant(3)}%</Trans>)
```
Comment (alfetopito) [inline src/components/CurrencyInputPanel/FiatValue.tsx:38]: What's translatable here? o.O

**Review by alfetopito — COMMENTED** [review]
"saving" first 90 files reviewed

File: src/components/TransactionConfirmationModal/index.tsx [hunk 4]
```diff
@@ -11,25 +12,25 @@ import { ButtonPrimary, ButtonLight } from '../Button'
...
+import { Trans } from '@lingui/macro'
 
 const Wrapper = styled.div`
   width: 100%;
+  padding: 1rem;
 `
-const Section = styled(AutoColumn)`
-  padding: 24px;
+const Section = styled(AutoColumn)<{ inline?: boolean }>`
+  padding: ${({ inline }) => (inline ? '0' : '0')};
```
Comment (alfetopito) [inline src/components/TransactionConfirmationModal/index.tsx:24]: WUT?

File: src/components/TransactionSettings/index.tsx [hunk 5]
```diff
@@ -85,165 +87,168 @@ const SlippageEmojiContainer = styled.span`
...
-    try {
-      const valueAsInt: number = Number.parseInt(value) * 60
-      if (!Number.isNaN(valueAsInt) && valueAsInt > 0) {
-        setDeadline(valueAsInt)
+    if (value.length === 0) {
+      setDeadline(DEFAULT_DEADLINE_FROM_NOW)
+    } else {
+      try {
+        const parsed: number = Math.floor(Number.parseFloat(value) * 60)
+        if (!Number.isInteger(parsed) || parsed < 60 || parsed > 180 * 60) {
```
Comment (alfetopito) [inline src/components/TransactionSettings/index.tsx:142]: To keep in mind.
Uni's allowed deadline is between 1 and 180 min.

File: src/components/swap/SwapModalHeader.tsx [hunk 6]
```diff
@@ -1,125 +1,185 @@
...
-    allowedSlippage
-  ])
-  const { priceImpactWithoutFee } = useMemo(() => computeTradePriceBreakdown(trade), [trade])
-  const priceImpactSeverity = warningSeverity(priceImpactWithoutFee)
-
   const theme = useContext(ThemeContext)
 
+  const [showInverted, setShowInverted] = useState<boolean>(false)
+
+  const fiatValueInput = useUSDCValue(trade.inputAmount)
```
Comment (alfetopito) [inline src/components/swap/SwapModalHeader.tsx:60]: Tcharaaaaan, called it! Using USDC

File: src/components/swap/SwapRoute.tsx [hunk 7]
```diff
@@ -1,26 +1,44 @@
...
 import { ThemeContext } from 'styled-components'
 import { TYPE } from 'theme'
-import { unwrappedToken } from 'utils/wrappedCurrency'
+import { unwrappedToken } from 'utils/unwrappedToken'
 
-export default memo(function SwapRoute({ trade }: { trade: Trade }) {
+function LabeledArrow({}: { fee: FeeAmount }) {
+  const theme = useContext(ThemeContext)
+
+  // todo: render the fee in the label
```
Comment (alfetopito) [inline src/components/swap/SwapRoute.tsx:14]: Is this our TODO?

File: src/connectors/index.ts [hunk 8]
```diff
@@ -3,58 +3,82 @@ import { InjectedConnector } from '@web3-react/injected-connector'
...
+const NETWORK_URLS: {
+  [chainId in SupportedChainId]: string
+} = {
+  [SupportedChainId.MAINNET]: `https://mainnet.infura.io/v3/${INFURA_KEY}`,
+  [SupportedChainId.RINKEBY]: `https://rinkeby.infura.io/v3/${INFURA_KEY}`,
+  [SupportedChainId.ROPSTEN]: `https://ropsten.infura.io/v3/${INFURA_KEY}`,
+  [SupportedChainId.GOERLI]: `https://goerli.infura.io/v3/${INFURA_KEY}`,
+  [SupportedChainId.KOVAN]: `https://kovan.infura.io/v3/${INFURA_KEY}`,
+  [SupportedChainId.ARBITRUM_KOVAN]: `https://kovan5.arbitrum.io/rpc`,
+  [SupportedChainId.ARBITRUM_ONE]: `https://arb1.arbitrum.io/rpc`,
```
Comment (alfetopito) [inline src/connectors/index.ts:31]: New networks, cool!

**Review by alfetopito — COMMENTED** [review]
Saving another 97 files (187 total so far).
Will start /custom/ next

File: src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx [hunk 9]
```diff
@@ -86,43 +100,49 @@ function TransactionSubmittedContent({
...
+            <Trans>Transaction Submitted</Trans>
           </Text>
           {chainId && hash && (
+            // <ExternalLink href={getExplorerLink(chainId, hash, ExplorerDataType.TRANSACTION)}>
             <ExternalLink href={getEtherscanLink(chainId, hash, 'transaction')}>
               <Text fontWeight={500} fontSize={14} color={theme.primary1}>
                 {/* View on Etherscan */}
-                {getExplorerLabel(chainId, hash, 'transaction')}
+                {/* <Trans>View on Explorer</Trans> */}
+                <Trans>{getExplorerLabel(chainId, hash, 'transaction')}</Trans>
```
Comment (alfetopito) [inline src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx:123]: I don't think it makes sense to translate this, as it won't be known at compile time.
Or will it?

File: src/custom/components/swap/EthWethWrap/index.tsx [hunk 10]
```diff
@@ -122,7 +123,9 @@ const ButtonWrapper = styled.div`
 
 const ErrorMessage = ({ error }: { error: Error }) => (
   <ErrorWrapper>
-    <strong>{error.message}</strong>
+    <strong>
+      <Trans>{error.message}</Trans>
```
Comment (alfetopito) [inline src/custom/components/swap/EthWethWrap/index.tsx:127]: Also don't think it'll be translatable at build time

File: src/custom/pages/App/AppMod.tsx [hunk 11]
```diff
@@ -77,24 +79,26 @@ const Marginer = styled.div`
...
-          {/* <TopLevelModals /> */}
-          <Web3ReactManager>
-            <Switch>
-              {props && props.children}
-              {/* <Route exact strict path="/swap" component={Swap} />
+    <ErrorBoundary>
+      <Suspense fallback={null}>
+        <Route component={GoogleAnalyticsReporter} />
+        <Route component={DarkModeQueryParamReader} />
+        <Route component={ApeModeQueryParamReader} />
```
Comment (alfetopito) [inline src/custom/pages/App/AppMod.tsx:88]: lol what?

File: src/custom/pages/Swap/SwapMod.tsx [hunk 12]
```diff
@@ -289,50 +337,59 @@ export default function Swap({
...
+  //   )
+  // }, [priceImpact, trade])
+
+  const isArgentWallet = useIsArgentWallet()
 
   // show approve flow when: no error on inputs, not approved or pending, or approved in current session
   // never show if price impact is above threshold in non expert mode
   const showApproveFlow =
+    // TODO: review this
+    !isArgentWallet &&
```
Comment (alfetopito) [inline src/custom/pages/Swap/SwapMod.tsx:388]: I think this has to do with token approval without a tx, or whatever that is.
The one where the token supports allowance natively, and I guess by this point only argent has that implemented?

File: src/custom/pages/Swap/SwapMod.tsx [hunk 13]
```diff
@@ -555,67 +703,120 @@ export default function Swap({
...
+                        <CurrencyLogo
+                          currency={currencies[Field.INPUT]}
+                          size={'20px'}
+                          style={{ marginRight: '8px', flexShrink: 0 }}
+                        />
+                        {/* we need to shorten this string on mobile */}
+                        {approvalState === ApprovalState.APPROVED || signatureState === UseERC20PermitState.SIGNED ? (
+                          <Trans>You can now trade {currencies[Field.INPUT]?.symbol}</Trans>
+                        ) : (
+                          <Trans>Allow the Uniswap Protocol to use your {currencies[Field.INPUT]?.symbol}</Trans>
```
Comment (alfetopito) [inline src/custom/pages/Swap/SwapMod.tsx:747]: TODO: remember to overwrite `Uniswap` here or in the index

File: src/custom/state/orders/consts.ts [hunk 14]
```diff
@@ -1,11 +1,11 @@
-import { ChainId } from '@uniswap/sdk'
+import { SupportedChainId as ChainId } from 'constants/chains'
 
 // TODO: fill contract deploymentblocks
 // to start checking for orders from that point
```
Comment (alfetopito) [inline src/custom/state/orders/consts.ts:4]:
```suggestion
```

File: src/custom/utils/xdai/hack.ts [hunk 15]
```diff
@@ -1,6 +1,7 @@
 // this file essentially provides all the overrides employed in uniswap-xdai-sdk fork
 // + logic for chainId switch to/from xDAI
```
Comment (alfetopito) [inline src/custom/utils/xdai/hack.ts:2]: Could this hack be removed maybe, now that Uni supports other chains?

For another PR, or course.

**Review by alfetopito — COMMENTED** [review]
Saving review up to file 322. Done with all the `/custom/` files.

David, thanks for jumping on the grenade for this one. Man, I'm sorry for you :sweat_smile: 

About the mod files and our custom changes in the respective indexes. I noticed that some are not in sync with their base files.
A follow up task I think is, compare against the base files once again to be sure nothing is lost on the way.

File: src/state/application/reducer.ts [hunk 16]
```diff
@@ -35,18 +35,18 @@ export default createReducer(initialState, builder =>
     .addCase(setOpenModal, (state, action) => {
       state.openModal = action.payload
     })
-    .addCase(addPopup, (state, { payload: { content, key, removeAfterMs = 15000 } }) => {
-      state.popupList = (key ? state.popupList.filter(popup => popup.key !== key) : state.popupList).concat([
+    .addCase(addPopup, (state, { payload: { content, key, removeAfterMs = 25000 } }) => {
```
Comment (alfetopito) [inline src/state/application/reducer.ts:38]: Longer pop up by default? :shrug:

**Review by alfetopito — COMMENTED** [review]
Saving the state at file 458.
Skipped a bunch of stuff that we don't use like pool, vote, governance, claim, etc

**Review by alfetopito — COMMENTED** [review]
Done with the file review!
Now on to some testing...

**Review by alfetopito — COMMENTED** [review]
Tested a few flows:

- Sell
- Buy
- Cancellation
- Sell ETH -> wrap

It works! All flows were good.

Style is of course completely fucked, but I know there's a pending PR for that so I won't talk about it.

Things to fix, likely not on this PR, given the size...

- Something is up with the state:
screenshot_2021-06-29_10-21-13
- "Disabled wallets" filter is not applied:
screenshot_2021-06-29_10-22-01
- WETH shows up as WETH9

Other than that, the app is taking much longer to start.

I'll leave a "soft" approval from the working point of view.
I also left several comments while reviewing the code.
Not all of them are addressable, some are just comments on uni's base code.

File: .github/workflows/integration-tests.yaml [hunk 17]
```diff
@@ -0,0 +1,50 @@
+name: Integration Tests
```
[the remark this answers is not in the document]
Comment (W3stside) [reply .github/workflows/integration-tests.yaml:1]: 👍

File: src/components/AccountDetails/Copy.tsx (hunk 1 above)
Comment (W3stside) [reply src/components/AccountDetails/Copy.tsx:2]: it's a babel bypass

File: src/components/CurrencyInputPanel/FiatValue.tsx (hunk 3 above)
Comment (W3stside) [reply src/components/CurrencyInputPanel/FiatValue.tsx:38]: lol no idea

File: src/components/TransactionConfirmationModal/index.tsx (hunk 4 above)
Comment (W3stside) [reply src/components/TransactionConfirmationModal/index.tsx:24]: unibois were tired i guess

File: src/components/swap/SwapRoute.tsx (hunk 7 above)
Comment (W3stside) [reply src/components/swap/SwapRoute.tsx:14]: nope

File: src/custom/components/AccountDetails/AccountDetailsMod.tsx [hunk 18]
```diff
@@ -219,199 +219,216 @@ export interface AccountDetailsProps {
...
-//         </LowerSection>
-//       ) : (
-//         <LowerSection>
-//           <TYPE.body color={theme.text1}>Your transactions will appear here...</TYPE.body>
-//         </LowerSection>
-//       )}
-//     </>
-//   )
-// }
+/*
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/AccountDetails/AccountDetailsMod.tsx:222]: it shows less changes and will show a line change within the block so imo better

File: src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx (hunk 9 above)
Comment (W3stside) [reply src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx:123]: yes, true

File: src/custom/components/swap/EthWethWrap/index.tsx (hunk 10 above)
Comment (W3stside) [reply src/custom/components/swap/EthWethWrap/index.tsx:127]: also true

File: src/custom/pages/App/AppMod.tsx (hunk 11 above)
Comment (alfetopito) [reply src/custom/pages/App/AppMod.tsx:88]: Found out what that is. Ape mode query param === expert mode

File: cypress/support/commands.js [hunk 19]
```diff
@@ -6,17 +6,23 @@
 
 import { JsonRpcProvider } from '@ethersproject/providers'
 import { Wallet } from '@ethersproject/wallet'
-import { _Eip1193Bridge } from '@ethersproject/experimental/lib/eip1193-bridge'
+import { Eip1193Bridge } from '@ethersproject/experimental/lib/eip1193-bridge'
 
-// never send real ether to this, obviously
-const PRIVATE_KEY_TEST_NEVER_USE = '0xad20c82497421e9784f18460ad2fe84f73569068e98e270b3e63743268af5763'
+// todo: figure out how env vars actually work in CI
+// const TEST_PRIVATE_KEY = Cypress.env('INTEGRATION_TEST_PRIVATE_KEY')
```
[the remark this answers is not in the document]
Comment (W3stside) [reply cypress/support/commands.js:12]: that's their comment :)

File: src/custom/components/AccountDetails/TransactionMod.tsx [hunk 20]
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import styled from 'styled-components'
 import { CheckCircle, Triangle } from 'react-feather'
 
-import { useActiveWeb3React } from 'hooks'
+import { useActiveWeb3React } from 'hooks/web3'
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/AccountDetails/TransactionMod.tsx:5]: added the comment of original code :)

File: src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx (hunk 9 above)
Comment (W3stside) [reply src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx:123]: but we can add it later via id's in the language files, i'll keep it around

File: src/custom/pages/Swap/SwapMod.tsx (hunk 12 above)
Comment (W3stside) [reply src/custom/pages/Swap/SwapMod.tsx:388]: yeah i need to look more into isArgentWallet

File: src/custom/pages/Swap/SwapMod.tsx (hunk 13 above)
Comment (W3stside) [reply src/custom/pages/Swap/SwapMod.tsx:747]: done in another PR 👍

File: src/theme/components.tsx [hunk 21]
```diff
@@ -197,6 +197,50 @@ export const UniTokenAnimated = styled.img`
   filter: drop-shadow(0px 2px 4px rgba(0, 0, 0, 0.15));
 `
 
+const ETHERSCAN_HOSTNAMES: { [hostname: string]: true } = {
+  'etherscan.io': true,
+  'ropsten.etherscan.io': true,
+  'rinkeby.etherscan.io': true,
+  'kovan.etherscan.io': true,
+  'goerli.etherscan.io': true,
+}
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/theme/components.tsx:206]: not sure but i will leave it there for now

File: src/utils/retry.ts [hunk 22]
```diff
@@ -10,6 +10,7 @@ function waitRandom(min: number, max: number): Promise<void> {
  * This error is thrown if the function is cancelled before completing
  */
 export class CancelledError extends Error {
+  public isCancelledError: true = true
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/utils/retry.ts:13]: haha Hum indeed

File: src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx (hunk 9 above)
Comment (alfetopito) [reply src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx:124]: What do you mean by via ids?

Comment (W3stside) [reply src/custom/components/TransactionConfirmationModal/TransactionConfirmationModalMod.tsx:124]: inside `src/locales` each locale file has `id`s set for translating. so i meant we can just add these translations later
e.g

```
#: src/components/AccountDetails/index.tsx
msgid "(clear all)"
msgstr "(clear all)"
```

File: src/custom/state/orders/middleware.ts [hunk 23]
```diff
@@ -26,9 +26,9 @@ const isFulfillOrderAction = isAnyOf(OrderActions.addPendingOrder, OrderActions.
 const isExpireOrdersAction = isAnyOf(OrderActions.expireOrdersBatch, OrderActions.expireOrder)
 const isCancelOrderAction = isAnyOf(OrderActions.cancelOrder, OrderActions.cancelOrdersBatch)
 
-// On each Pending, Expired, Fulfilled, Cancelled order action
-// a corresponding Popup action is dispatched
-export const popupMiddleware: Middleware<{}, AppState> = store => next => action => {
+// on each Pending, Expired, Fulfilled order action
+// a corresponsing Popup action is dispatched
+export const popupMiddleware: Middleware<Record<string, unknown>, AppState> = (store) => (next) => (action) => {
```
Comment (alfetopito) [inline src/custom/state/orders/middleware.ts:31]: Nit: comment typo fixes are being reverted

**Comment by alfetopito** [comment]
Also the build on this PR leads to an empty page.
Locally it's fine though.
I'm wonder whether we need to do something regarding the CI?

Also also, (probably in another PR), integration need fixing.

**Pull request merged**